### PR TITLE
Add an Alloy spec of FFG

### PR DESCRIPTION
**Pull request by konnov**
This is another attempt at pinpointing the bottlenecks in model checking the FFG spec. In theory, Alloy seems to be better suited for graph-theoretic problems. We have to collect the experimental data. In the preliminary results, Alloy also works for very small parameters.

### CODE REVIEW

File: abstract-spec/ffg.als [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	all v: FfgVote | {
+		isAncestorDescendant[v.source.block, v.target.block]
+		v.source.slot < v.target.slot
+	}
+}
+
+fact checkpointsAreValid {
+	all c: Checkpoint | {
+		(c.slot = 0 and c.block = GenesisBlock) or c.slot > c.block.slot
+	}
```
Comment (thpani) [inline abstract-spec/ffg.als:94]: Question for @banhday and @saltiniroberto:

This constraint is in the paper, but not in the Python spec.
Should we keep it?

File: abstract-spec/ffg.als [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+run someFinalizedCheckpoint { some c: Checkpoint | isFinalized[c] and c.slot != 0 }
+	for 10 but 6 Block, 6 Checkpoint, 12 Vote, 5 int
+run someConflicting { some b1: Block, b2: Block | areConflictingBlocks[b1, b2] }
+	for 6 but 3 Block, 6 Checkpoint, 5 int
+run someSlashable { some slashableNodes }
+	for 6 but 3 Block, 6 Checkpoint, 5 int
+run disagreement for 10 but 5 Block, 6 Checkpoint, 20 Vote, 5 int
+run accountableSafety for 10 but 5 Block, 6 Checkpoint, 20 Vote, 5 int
+run noAccountableSafety for 10 but 5 Block, 6 Checkpoint, 20 Vote, 5 int
+run { some b: Block | b != GenesisBlock } for 6 but 5 Block, 5 Checkpoint, 5 int
```
Comment (thpani) [inline abstract-spec/ffg.als:208]:
```suggestion
```

**Review by thpani — APPROVED** [review]
Generally looks good. I would like confirmation on the checkpoint constraint below:

**Comment by konnov** [comment]
Merging this spec, since we are not introducing any changes to it any longer. The only question is factored out in the issue #54

**Pull request merged**